### DISCUSSION TITLE
Replacing Dispatcher from UWP

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [microsoft/microsoft-ui-xaml#7584] [Q&A] [answered]
I'm looking at the Midi sample for UWP, it uses CoreDispatcher. How can I replace that with DispatcherQueue? It doesn't have a RunAsync method. So I don't understand how DispatcherQueue can replace it. Can I just get rid of the awaits in these usage of CoreDispatcher and make the methods synchronous?

```cs
// UWP Code
        private async void DeviceWatcher_Added(DeviceWatcher sender, DeviceInformation args)
        {
            // If all devices have been enumerated
            if (this.enumerationCompleted)
            {
                await coreDispatcher.RunAsync(CoreDispatcherPriority.High, () =>
                {
                    // Update the device list
                    UpdateDevices();
                });
            }
        }
```

**Comment by Panda-Sharp** [comment, marked as answer]
The toolkit has an extension for that
https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/rel/7.1.2/Microsoft.Toolkit.Uwp/Extensions/DispatcherQueueExtensions.cs

and here you can find a sample
https://github.com/CommunityToolkit/WindowsCommunityToolkit/tree/rel/7.1.2/Microsoft.Toolkit.Uwp.SampleApp/SamplePages/DispatcherQueueHelper

**Comment by castorix** [comment]
You can see the Midi sample I had posted : [WinUI3_Midi](https://github.com/castorix/WinUI3_Midi)
(in task _async System.Threading.Tasks.Task MyTask()_)

**Reply by Gavin-Williams** [reply]
Awesome, thanks. I got rid of the async code just to get my code working today. I have to give it some thought tomorrow, look at the toolkit extensions and I'll look at your sample. I'm not sure yet why device connection even needs to be async - I need to have a closer look at that part of it. Work out what's happening, and on what threads. I guess any kind of device negotiation may be slow? Anyway, I did connect to a midi-keyboard and played some notes, controlled it in other ways - which is pretty entertaining for how little code was needed. Tomorrow I will look closer and also work out how to input midi signals. I'm really impressed with how straight forward midi is to use. I don't know if it's always been this way. But it seems like a good system to communicate with musical instruments.
